File: Python/783. Minimum Distance Between BST Nodes/minDiffInBST.py
#Time complexity: O(n)
#Space complexity: O(n)

"""Description

Given the root of a Binary Search Tree (BST), return the minimum difference
between the values of any two different nodes in the tree.

Example:
    Input: root = [4,2,6,1,3]
    Output: 1

"""


class Solution:

    def minDiffInBST(self, root):
        """Implementation of problem solution.

            :param root: reference to root node of BST tree.
            :return: minimum difference between the values of
                    any two different nodes in the tree.
        """
        MAX_VALUE = 1e5 + 1
        nodes_value_list = []
        build_list_by_traverse_tree_inOrder(root, nodes_value_list)
        min_diff = MAX_VALUE
        for i in range(len(nodes_value_list) - 1, 0, -1):
            if (diff := nodes_value_list[i] - nodes_value_list[i - 1]) < min_diff:
                min_diff = diff
        return min_diff


def build_list_by_traverse_tree_inOrder(node, nodes_value_list):
    """Recurrence function for building list of nodes value in order.

        :param node: reference to node of BST tree.
        :param nodes_value_list: reference to list of nodes values.
    """
    if not node:
        return
    build_list_by_traverse_tree_inOrder(node.left, nodes_value_list)
    nodes_value_list.append(node.val)
    build_list_by_traverse_tree_inOrder(node.right, nodes_value_list)

File: Solutions/783. Minimum Distance Between BST Nodes/C++/minDiffInBST.cpp
//Time complexity: O(n)
//Space complexity: O(n)

/*Description

Given the root of a Binary Search Tree (BST), return the minimum difference
between the values of any two different nodes in the tree.

Example:
    Input: root = [4,2,6,1,3]
    Output: 1

*/

const int MAX_NODES_DIFF = 1e5 + 1;


class Solution {
public:
    int minDiffInBST(TreeNode* root) {
        std::vector<int> bstNodesVal;
        buildNodeValArrByTraverseTreeInOrder(root, bstNodesVal);
        return getMinInBst(bstNodesVal);
    }


    void buildNodeValArrByTraverseTreeInOrder(TreeNode* node, std::vector<int>& nodesValVector)
    {
        if (!node)
        {
            return;
        }
        buildNodeValArrByTraverseTreeInOrder(node->left, nodesValVector);
        nodesValVector.push_back(node->val);
        buildNodeValArrByTraverseTreeInOrder(node->right, nodesValVector);
    }

    int getMinInBst(std::vector<int>& nodesValVector)
    {
        int minDiffVal = MAX_NODES_DIFF;
        int greaterNodeVal, lowerNodeVal, currNodeVal;
        greaterNodeVal = nodesValVector.back();
        nodesValVector.pop_back();
        while (!nodesValVector.empty())
        {
            lowerNodeVal = nodesValVector.back();
            currNodeVal = greaterNodeVal - lowerNodeVal;
            minDiffVal = currNodeVal < minDiffVal ? currNodeVal : minDiffVal;
            greaterNodeVal = lowerNodeVal;
            nodesValVector.pop_back();
        }
        return minDiffVal;
    }
};